modules/videoio/cmake/detect_ffmpeg.cmake

```diff
@@ -19,10 +19,10 @@ if(NOT HAVE_FFMPEG AND WIN32 AND NOT ARM AND NOT OPENCV_FFMPEG_SKIP_DOWNLOAD)
   endif()
 endif()
 
-set(_required_ffmpeg_libraries libavcodec libavformat libavutil libswscale)
+set(_required_ffmpeg_libraries libavcodec libavformat libavutil libswscale libavdevice)
 set(_used_ffmpeg_libraries ${_required_ffmpeg_libraries})
 if(NOT HAVE_FFMPEG AND PKG_CONFIG_FOUND)
-  ocv_check_modules(FFMPEG libavcodec libavformat libavutil libswscale)
+  ocv_check_modules(FFMPEG libavcodec libavformat libavutil libswscale libavdevice)
```

Comment:
Need to check if `libavdevice` module is optional or not in `FFmpeg` and `libav` packages.
If it is optional, then it should be added as optional into OpenCV (to avoid breaking of existed user configurations)

---

**Update**:

Looks like some configurations may miss libavdevice:

E.g., Ubuntu 20.04:

```
# ll /usr/lib/x86_64-linux-gnu/pkgconfig/{libav*,libswscale*}
-rw-r--r-- 1 root root 742 May 18 21:24 /usr/lib/x86_64-linux-gnu/pkgconfig/libavcodec.pc
-rw-r--r-- 1 root root 472 May 18 21:24 /usr/lib/x86_64-linux-gnu/pkgconfig/libavformat.pc
-rw-r--r-- 1 root root 386 May 18 21:24 /usr/lib/x86_64-linux-gnu/pkgconfig/libavutil.pc
-rw-r--r-- 1 root root 322 May 18 21:24 /usr/lib/x86_64-linux-gnu/pkgconfig/libswscale.pc
```

Installed through `apt-get install libavcodec-dev libavformat-dev libswscale-dev`

---

Optional dependency means these steps:
- CMake cache variable to control behavior - `OPENCV_FFMPEG_ENABLE_LIBAVDEVICE`
- default value is OFF (to avoid breaking of user code)
- add compiler macro to pass into C++ code if libavdevice enabled and should be used (e.g., `OPENCV_FFMPEG_LIBAVDEVICE=1`)
- related C++ code should be guarded by `#ifdef OPENCV_FFMPEG_LIBAVDEVICE`

Reply:
Thanks! makes sense. I am on it.

Reply:
Just about the first step. Shouldn't i do it like [that](https://github.com/opencv/opencv/blob/1293e99ad8e97a3ee1968787d77111a335c47a8f/CMakeLists.txt#L1442-L1446)?

Reply:
and then like [that](https://github.com/opencv/opencv/blob/17234f82d025e3bbfbf611089637e5aa2038e7b8/modules/videoio/cmake/detect_ffmpeg.cmake#L27-L31)?

Reply:
I think i figured it out right: eb7bc8ef3c6f45642efd5c0e8bf23fb400814e47